Sidetrack: Fix Error handling and UI

MegFord opened the pull request:
Fixes outstanding issues in Sidetrack Toolbox:
Checking for <8 instructions
Showing instructions error on right line
Validation of the level editor code
Fix Topic title shadow

https://phabricator.endlessm.com/T26109

ptomato commented on src/RileyMaze/userFunction.js:
```diff
@@ -198,7 +186,10 @@ ${code}
         const scope = getScope();
-        scope.riley.queue.length = 0;
+
+        if (scope.hasOwnProperty('riley'))
```
This seems like it should be moved into the Instructions topic's `getScope` function, since the `riley` object only applies there. I'd like to avoid special-casing things in the UserFunction class because I think the common user function code should eventually be factored out of Lightspeed and Sidetrack.

ptomato commented on src/RileyMaze/userFunction.js:
```diff
@@ -240,6 +232,22 @@ ${code}
             return;
         }
 
+        if (scope.hasOwnProperty('riley') && scope.riley.queue.length !== MAX_QUEUE_LEN) {
```
Likewise I'd prefer to avoid this special-case as well, if possible, because `if (scope.hasOwnProperty('riley'))` is basically saying `if (this is the "instructions" user function)`. Maybe it would work to add a `validate()` function to the descriptions in `USER_FUNCTIONS` similar to `getScope()`, and put this code there?

ptomato commented on src/RileyMaze/userFunction.js:
```diff
@@ -240,6 +232,22 @@ ${code}
 
+        if (scope.hasOwnProperty('riley') && scope.riley.queue.length !== MAX_QUEUE_LEN) {
+            // Add lines for the function name
+            let lineNum = MAX_QUEUE_LEN + 2;
```
I don't feel very comfortable with guessing the line number, as it can easily lead to unintentionally confusing error messages when the player does something we don't expect:
![Screenshot from 2019-04-26 12-22-46](https://user-images.githubusercontent.com/459371/56831681-8fa87400-681e-11e9-990b-6b2bde00444b.png)
If I'm reading the code correctly, you removed the error-throwing code from the methods of the Riley object because of https://phabricator.endlessm.com/T26436 ? I think I would prefer here to reinstate those, and instead fix the underlying bug in a separate PR. Throwing the error seems to be the correct way to get the error message at the correct line, but it just happens to not work at the moment...

MegFord replied:
Yeah, I agree, thanks!

ptomato commented on src/RileyMaze/userFunction.js:
```diff
@@ -240,6 +232,22 @@ ${code}
+            // Add lines for the function name
+            let lineNum = MAX_QUEUE_LEN + 2;
+            if (scope.riley.queue.length < MAX_QUEUE_LEN)
+                lineNum = scope.riley.queue.length + 1;
```
IIRC in the discussion on Randy's spec, we landed on showing the error on the last line of the code view in the case of too few instructions.

ptomato reviewed (changes requested):
The UI widgets in the units panel look much better now!

MegFord committed:
Sidetrack: Fix Error handling

Fixes outstanding issues in Sidetrack Toolbox:
Checking for <8 instructions
Showing instructions error on right line
Validation of the level editor code

https://phabricator.endlessm.com/T26109

ptomato reviewed (approved):
Great, thanks!
